### PR TITLE
gnrc_ipv6_nib/nc: remove from _next_removable on remove

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
The `_next_removable` list manages the cache-out of the neighbor cache. However, when a neighbor cache entry is removed, it is not removed from that list, which may lead to a segmentation fault when that list is accessed, since the whole entry (including its list pointer) is zeroed
after removal.

With this change the entry is removed from that list accordingly before the zeroing happens.

### Testing procedure
~~This is hard to unittests, since the removal needs to happen *while* another entry is added (so this is a race condition though I'm not 100% clear how this happens, since both operations are supposed to run in the `gnrc_ipv6` thread). I used https://gist.github.com/miri64/fac4df86be36f0a65d9bdb4d2f09d5c7#file-03-4-test-sh and obseverd `tap0` (you can do that by typing `tmux select-window -t tap0; tmux a` in another terminal). Without this PR a segfault will happen here~~

~~https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c#L153~~

~~With this PR this segfault does not happen.~~

There is a unittest for this edge case in this PR ;-).

### CODE REVIEW

**Comment by miri64** [comment]
Because of

> (so this is a race condition though I'm not 100% clear how this happens, since both operations are supposed to run in the `gnrc_ipv6` thread)

I might have opened this PR prematurely.

**Comment by miri64** [comment]
Ok, I think I got it now! The state corruption happens definitely within the thread, but not due to a remove happening just as something is added to the neighbor cache, but much earlier. It does not lead to a problem immediately due to a quirk of `clist` where the first element is not `list` or `list.next`, but `list.next->next`, so that most of its operations can be done in *O(1)*: `list` is just the entry point into the list, so it never appears in it itself, `list.next` is the last element of the list, so things can be appended in *O(1)* and `list.next->next` is the first.

It's pretty hard to explain, but when the last two **elements** are by change removed directly after each other when the NC is full (which introduces NULL pointers at the last and first **position**) and then two new NC entries are added (which have the same pointer as the removed ones), the NULL pointer at the **position** is first moved to the second **position** and then kept there, due to 

https://github.com/RIOT-OS/RIOT/blob/e78069b3163ec61ec5b691c1eefa5d68309b3263/core/include/clist.h#L115

When another entry is tried to be added, the cache-out mechanism starts running (since the NC is full again) and tries to pop the first **element** (which is the NULL pointer) and it crashes.

When just the **last** element is removed the situation "fixes" itself, since the entry is still referred to by the first position, so re-adding it just leads to a loop of one (breaking the list, but not the system ;-)).

I provided a unittest to reproduce and illustrate the crash (2451fb36f8d5abea2cbb10cc61d60f597a0518e2 while writing this message).

**Comment by miri64** [comment]
> When just the **last** element is removed the situation "fixes" itself, since the entry is still referred to by the first position, so re-adding it just leads to a loop of one (breaking the list, but not the system ;-)).

@aabadie I think some of the leaks you were seeing in https://github.com/RIOT-OS/Release-Specs/issues/98 were in fact caused by this.

File: tests/unittests/tests-gnrc_ipv6_nib/tests-gnrc_ipv6_nib-internal.c [hunk 1]
```diff
@@ -302,6 +302,10 @@ static void test_nib_iter__three_elem_middle_removed(void)
     addr.u64[1].u64++;
     TEST_ASSERT_NOT_NULL((node3 = _nib_onl_alloc(&addr, IFACE)));
     node3->mode = _DRL;
+    /* cppcheck-suppress redundantAssignment
+     * (reason: we assigned _FT before so _nib_onl_alloc would recognize node2
+     *          as used, now we want to clear it, so we need to set it to
+     *          _EMPTY... we are testing internals of data structures here) */
```
Comment (miri64) [inline tests/unittests/tests-gnrc_ipv6_nib/tests-gnrc_ipv6_nib-internal.c]: Unrelated but needed to pass CI.

**Comment by aabadie** [comment]
I tested with PR with 10 nodes pinging a single one and get a non empty pktbuf at the end. See the content [here](https://gist.github.com/aabadie/59d5a58392eab9be73bde47b7e278e41).

No idea if this PR is supposed to fix this or not. And also no idea what the content of the pktbuf means.

**Comment by miri64** [comment]
This PR is supposed to fix a segfault I encountered, so due to #10978 alone there will still be leaks of course.

**Comment by miri64** [comment]
See testing procedures for how to test ;-).

**Comment by aabadie** [comment]
> See testing procedures for how to test ;-)

Ok, I ran the unittests on native and they are passing. I guess I can ACK ?

**Comment by aabadie** [comment]
Please provide a backport

**Comment by miri64** [comment]
Backport provided in #10993

**Pull request merged**